Mergining in Guidance fixes and functional controller tests

briri opened the pull request:
Fixes for #204 and #221. This code also includes the majority of the functional controller tests outlined in #212.

All tests should pass before accepting future PRs after this is merged! 

Note that it looked like there were some missing pieces to the plan pages. For example the `views/plans/_answer_form_ro.html.erb` was missing so I just added a stub in. If you have a more accurate copy @raycarrick-ed, please replace this one with yours. The registratations controller also had some typos in it that were preventing some of the various results to fail. 

Also be aware that I touched almost all of the controllers to update/correct how they were processing error conditions. Please pay close attention to those when you're reviewing.

jollopre commented on app/policies/plan_policy.rb:
```diff
@@ -28,17 +28,25 @@ def export?
-    @plan.readable_by?(@user.id)
-  end
+# TODO: There is no route to this function on the controller
+#  def show_export?
```
Uncomment this, otherwise the export page will not be displayed. See the route here https://github.com/DigitalCurationCentre/roadmap/blob/development/config/routes.rb#L201

briri replied:
ah ok. missed that one

raycarrick-ed commented on app/controllers/guidance_groups_controller.rb:
```diff
@@ -66,7 +66,7 @@ def admin_update_publish
       redirect_to admin_index_guidance_path(params[:guidance_group]), notice: _('Guidance group was successfully updated.')
     else
-      render action: "edit"
+      redirect_to admin_index_guidance_path(@guidance_group), notice: generate_error_notice(@guidance_group)
```
We should adopt this as general pattern. Much better than what we've been working with. It might be worth generalising the 'Unable to save your changes' message e.g under plan creation we could do the same but might want 'Unable to Create Plan'. Add as a param maybe?

briri replied:
Yes that's a good idea

briri commented on app/controllers/phases_controller.rb:
```diff
@@ -176,7 +179,8 @@ def admin_update
       redirect_to admin_show_phase_path(@phase), notice: _('Information was successfully updated.')
     else
-      render action: "admin_show"
+      # Redirecting here because the method loads a lot of stuff. We may lose whatever the user had entered though 
```
that's a good point. The redirect would likely lose whatever the user had entered. I'll switch them to renders

briri commented on app/controllers/plans_controller.rb:
```diff
@@ -17,121 +17,98 @@ def index
+    authorize Plan
+    
+    @plan = Plan.new
+    authorize @plan
```
Ah. good catch. No need for the first one here

briri commented on app/models/guidance.rb:
```diff
@@ -13,12 +13,10 @@ class Guidance < ActiveRecord::Base
   # Associations
   belongs_to :guidance_group
-#  belongs_to :question
+  belongs_to :question
```
yes. I thought I had it commented out but may have added it back in prior to removing those dropdowns from the guidance pages. I'll remove it.

briri commented:
Thanks for reviewing. I'll make the changes you have suggested.

briri commented:
Ok @raycarrick-ed and @jollopre. I made some changes based on your feedback:
- Uncommented show_export method in `plan_policy.rb` and `plan_controller.rb`
- broke the `generate_error_notice` method into 3 separate methods. One for each transaction type - create/update/destroy. Also updated it to require that you provide an object name since its not always equivalent to the class name (e.g. User vs Profile)
- Switched all save failure routes in controllers from redirect to render so that user does not lose the information they have already entered on the form
- Fixed tests to work with new error messaging and routing
